From 1af5cc59b0ec5c9af486e2df49c3023d39e69f0c Mon Sep 17 00:00:00 2001
From: Jernej Skrabec <jernej.skrabec@gmail.com>
Date: Fri, 22 May 2026 16:37:59 +0000
Subject: [PATCH] dt-bindings: display: Add A523 display engine compatibles

Allwinner A523/A527/T527 contains a Display Engine 3.5. It features a
single mixer, multiple TCONs (two TV TCONs and three LCD TCONs), and
HDMI / eDP / LVDS / DSI outputs. Add compatibles for the display
pipeline blocks.

The DE 3.5 mixer keeps the DE 3.3 register layout, so it reuses the
H616 mixer node binding. The TCON-TV variant matches the H616 one as
well. The TCON-TOP / display-if-top has a different layout: only one
mixer input (so no port@2/port@3 required) and no TVE clock.

Signed-off-by: Jernej Skrabec <jernej.skrabec@gmail.com>
---
 .../bus/allwinner,sun50i-a64-de2.yaml         | 18 ++++-
 .../clock/allwinner,sun8i-a83t-de2-clk.yaml   |  3 +
 .../allwinner,sun4i-a10-display-engine.yaml   |  4 +-
 .../display/allwinner,sun4i-a10-tcon.yaml     |  9 +++
 .../allwinner,sun50i-h616-de33-planes.yaml    | 10 ++-
 .../allwinner,sun8i-a83t-de2-mixer.yaml       |  2 +
 .../display/allwinner,sun8i-a83t-dw-hdmi.yaml |  5 ++
 .../allwinner,sun8i-a83t-hdmi-phy.yaml        |  1 +
 .../display/allwinner,sun8i-r40-tcon-top.yaml | 80 +++++++++++++++++--
 9 files changed, 120 insertions(+), 12 deletions(-)

diff --git a/Documentation/devicetree/bindings/bus/allwinner,sun50i-a64-de2.yaml b/Documentation/devicetree/bindings/bus/allwinner,sun50i-a64-de2.yaml
index f2ecdcd818fe..58606248f23e 100644
--- a/Documentation/devicetree/bindings/bus/allwinner,sun50i-a64-de2.yaml
+++ b/Documentation/devicetree/bindings/bus/allwinner,sun50i-a64-de2.yaml
@@ -29,6 +29,9 @@ properties:
       - items:
           - const: allwinner,sun50i-h616-de33
           - const: allwinner,sun50i-a64-de2
+      - items:
+          - const: allwinner,sun55i-a523-de35
+          - const: allwinner,sun50i-a64-de2
 
   reg:
     maxItems: 1
@@ -52,7 +55,8 @@ patternProperties:
     additionalProperties: true
     properties:
       reg:
-        maxItems: 1
+        minItems: 1
+        maxItems: 2
 
     required:
       - reg
@@ -63,7 +67,17 @@ required:
   - "#address-cells"
   - "#size-cells"
   - ranges
-  - allwinner,sram
+
+allOf:
+  - if:
+      not:
+        properties:
+          compatible:
+            contains:
+              const: allwinner,sun55i-a523-de35
+    then:
+      required:
+        - allwinner,sram
 
 additionalProperties: false
 
diff --git a/Documentation/devicetree/bindings/clock/allwinner,sun8i-a83t-de2-clk.yaml b/Documentation/devicetree/bindings/clock/allwinner,sun8i-a83t-de2-clk.yaml
index 7fcd55d468d4..4f9edc268cf7 100644
--- a/Documentation/devicetree/bindings/clock/allwinner,sun8i-a83t-de2-clk.yaml
+++ b/Documentation/devicetree/bindings/clock/allwinner,sun8i-a83t-de2-clk.yaml
@@ -26,6 +26,9 @@ properties:
       - const: allwinner,sun50i-h5-de2-clk
       - const: allwinner,sun50i-h6-de3-clk
       - const: allwinner,sun50i-h616-de33-clk
+      - items:
+          - const: allwinner,sun55i-a523-de35-clk
+          - const: allwinner,sun50i-h616-de33-clk
       - items:
           - const: allwinner,sun8i-r40-de2-clk
           - const: allwinner,sun8i-h3-de2-clk
diff --git a/Documentation/devicetree/bindings/display/allwinner,sun4i-a10-display-engine.yaml b/Documentation/devicetree/bindings/display/allwinner,sun4i-a10-display-engine.yaml
index 6e26f184ab24..253590c3f08f 100644
--- a/Documentation/devicetree/bindings/display/allwinner,sun4i-a10-display-engine.yaml
+++ b/Documentation/devicetree/bindings/display/allwinner,sun4i-a10-display-engine.yaml
@@ -66,6 +66,7 @@ properties:
       - allwinner,sun50i-a64-display-engine
       - allwinner,sun50i-h6-display-engine
       - allwinner,sun50i-h616-display-engine
+      - allwinner,sun55i-a523-display-engine
 
   allwinner,pipelines:
     $ref: /schemas/types.yaml#/definitions/phandle-array
@@ -75,7 +76,7 @@ properties:
       maxItems: 1
     description: |
       Available display engine frontends (DE 1.0) or mixers (DE
-      2.0/3.0/3.3) available.
+      2.0/3.0/3.3/3.5) available.
 
 required:
   - compatible
@@ -98,6 +99,7 @@ if:
           - allwinner,sun20i-d1-display-engine
           - allwinner,sun50i-a64-display-engine
           - allwinner,sun50i-h616-display-engine
+          - allwinner,sun55i-a523-display-engine
 
 then:
   properties:
diff --git a/Documentation/devicetree/bindings/display/allwinner,sun4i-a10-tcon.yaml b/Documentation/devicetree/bindings/display/allwinner,sun4i-a10-tcon.yaml
index eb8d1fd9d8a3..f79b04f79153 100644
--- a/Documentation/devicetree/bindings/display/allwinner,sun4i-a10-tcon.yaml
+++ b/Documentation/devicetree/bindings/display/allwinner,sun4i-a10-tcon.yaml
@@ -38,6 +38,11 @@ properties:
       - const: allwinner,sun20i-d1-tcon-tv
       - const: allwinner,sun50i-h616-tcon-tv
 
+      - items:
+          - enum:
+              - allwinner,sun55i-a523-tcon-tv
+          - const: allwinner,sun50i-h616-tcon-tv
+
       - items:
           - enum:
               - allwinner,sun7i-a20-tcon0
@@ -63,6 +68,7 @@ properties:
       - items:
           - enum:
               - allwinner,sun50i-h616-tcon-lcd
+              - allwinner,sun55i-a523-tcon-lcd
           - const: allwinner,sun8i-r40-tcon-lcd
 
   reg:
@@ -122,6 +128,9 @@ properties:
           - const: edp
           - const: lvds
 
+  power-domains:
+    maxItems: 1
+
   ports:
     $ref: /schemas/graph.yaml#/properties/ports
 
diff --git a/Documentation/devicetree/bindings/display/allwinner,sun50i-h616-de33-planes.yaml b/Documentation/devicetree/bindings/display/allwinner,sun50i-h616-de33-planes.yaml
index 0a07cb8a5f79..e0afd78f732d 100644
--- a/Documentation/devicetree/bindings/display/allwinner,sun50i-h616-de33-planes.yaml
+++ b/Documentation/devicetree/bindings/display/allwinner,sun50i-h616-de33-planes.yaml
@@ -4,20 +4,22 @@
 $id: http://devicetree.org/schemas/display/allwinner,sun50i-h616-de33-planes.yaml#
 $schema: http://devicetree.org/meta-schemas/core.yaml#
 
-title: Allwinner H616 Display Engine 3.3 planes
+title: Allwinner H616 Display Engine 3.3 / A523 Display Engine 3.5 planes
 
 maintainers:
   - Jernej Skrabec <jernej.skrabec@gmail.com>
 
 description:
-  Display Engine 3.3 planes are independent of mixers, contrary to
-  previous generations of Display Engine. Planes can be assigned to
-  mixers independently and even dynamically during runtime.
+  Display Engine 3.3 (and Display Engine 3.5) planes are independent of
+  mixers, contrary to previous generations of Display Engine. Planes can
+  be assigned to mixers independently and even dynamically during
+  runtime.
 
 properties:
   compatible:
     enum:
       - allwinner,sun50i-h616-de33-planes
+      - allwinner,sun55i-a523-de35-planes
 
   reg:
     maxItems: 1
diff --git a/Documentation/devicetree/bindings/display/allwinner,sun8i-a83t-de2-mixer.yaml b/Documentation/devicetree/bindings/display/allwinner,sun8i-a83t-de2-mixer.yaml
index f65682449c95..7e285ea0dab0 100644
--- a/Documentation/devicetree/bindings/display/allwinner,sun8i-a83t-de2-mixer.yaml
+++ b/Documentation/devicetree/bindings/display/allwinner,sun8i-a83t-de2-mixer.yaml
@@ -26,6 +26,7 @@ properties:
       - allwinner,sun50i-a64-de2-mixer-1
       - allwinner,sun50i-h6-de3-mixer-0
       - allwinner,sun50i-h616-de33-mixer-0
+      - allwinner,sun55i-a523-de35-mixer-0
 
   reg: true
 
@@ -75,6 +76,7 @@ allOf:
           contains:
             enum:
               - allwinner,sun50i-h616-de33-mixer-0
+              - allwinner,sun55i-a523-de35-mixer-0
     then:
       properties:
         reg:
diff --git a/Documentation/devicetree/bindings/display/allwinner,sun8i-a83t-dw-hdmi.yaml b/Documentation/devicetree/bindings/display/allwinner,sun8i-a83t-dw-hdmi.yaml
index 44cc5c2beb04..b113d6c04ced 100644
--- a/Documentation/devicetree/bindings/display/allwinner,sun8i-a83t-dw-hdmi.yaml
+++ b/Documentation/devicetree/bindings/display/allwinner,sun8i-a83t-dw-hdmi.yaml
@@ -32,6 +32,8 @@ properties:
           - const: allwinner,sun50i-h616-dw-hdmi
           - const: allwinner,sun50i-h6-dw-hdmi
 
+      - const: allwinner,sun55i-a523-dw-hdmi
+
       - items:
           - enum:
               - allwinner,sun8i-h3-dw-hdmi
@@ -92,6 +94,9 @@ properties:
     description:
       The VCC power supply of the controller
 
+  power-domains:
+    maxItems: 1
+
   ports:
     $ref: /schemas/graph.yaml#/properties/ports
 
diff --git a/Documentation/devicetree/bindings/display/allwinner,sun8i-a83t-hdmi-phy.yaml b/Documentation/devicetree/bindings/display/allwinner,sun8i-a83t-hdmi-phy.yaml
index fc88c5e2a3f6..e9e47df4f707 100644
--- a/Documentation/devicetree/bindings/display/allwinner,sun8i-a83t-hdmi-phy.yaml
+++ b/Documentation/devicetree/bindings/display/allwinner,sun8i-a83t-hdmi-phy.yaml
@@ -22,6 +22,7 @@ properties:
       - allwinner,sun50i-a64-hdmi-phy
       - allwinner,sun50i-h6-hdmi-phy
       - allwinner,sun50i-h616-hdmi-phy
+      - allwinner,sun55i-a523-hdmi-phy
 
   reg:
     maxItems: 1
diff --git a/Documentation/devicetree/bindings/display/allwinner,sun8i-r40-tcon-top.yaml b/Documentation/devicetree/bindings/display/allwinner,sun8i-r40-tcon-top.yaml
index 41324e1376ee..77f224fab007 100644
--- a/Documentation/devicetree/bindings/display/allwinner,sun8i-r40-tcon-top.yaml
+++ b/Documentation/devicetree/bindings/display/allwinner,sun8i-r40-tcon-top.yaml
@@ -44,16 +44,18 @@ properties:
       - allwinner,sun20i-d1-tcon-top
       - allwinner,sun50i-h6-tcon-top
       - allwinner,sun50i-h616-tcon-top
+      - allwinner,sun55i-a523-tcon-top
+      - allwinner,sun55i-a523-tcon-top1
 
   reg:
     maxItems: 1
 
   clocks:
-    minItems: 2
+    minItems: 1
     maxItems: 6
 
   clock-names:
-    minItems: 2
+    minItems: 1
     maxItems: 6
 
   clock-output-names:
@@ -63,6 +65,9 @@ properties:
   resets:
     maxItems: 1
 
+  power-domains:
+    maxItems: 1
+
   ports:
     $ref: /schemas/graph.yaml#/properties/ports
 
@@ -104,18 +109,27 @@ properties:
       - port@5
 
 required:
-  - "#clock-cells"
   - compatible
   - reg
   - clocks
   - clock-names
-  - clock-output-names
   - resets
-  - ports
 
 additionalProperties: false
 
 allOf:
+  - if:
+      not:
+        properties:
+          compatible:
+            contains:
+              const: allwinner,sun55i-a523-tcon-top1
+    then:
+      required:
+        - "#clock-cells"
+        - clock-output-names
+        - ports
+
   - if:
       properties:
         compatible:
@@ -234,6 +248,62 @@ allOf:
             - port@2
             - port@3
 
+  - if:
+      properties:
+        compatible:
+          contains:
+            const: allwinner,sun55i-a523-tcon-top
+
+    then:
+      properties:
+        clocks:
+          items:
+            - description: The TCON TOP interface clock
+            - description: The TCON TOP TV0 clock
+            - description: The TCON TOP TV1 clock
+            - description: The TCON TOP MIPI DSI clock
+
+        clock-names:
+          items:
+            - const: bus
+            - const: tcon-tv0
+            - const: tcon-tv1
+            - const: dsi
+
+        clock-output-names:
+          items:
+            - description: TCON TV0 output clock name
+            - description: TCON TV1 output clock name
+            - description: DSI output clock name
+
+        ports:
+          required:
+            - port@2
+            - port@3
+
+  - if:
+      properties:
+        compatible:
+          contains:
+            const: allwinner,sun55i-a523-tcon-top1
+
+    then:
+      properties:
+        clocks:
+          minItems: 1
+          items:
+            - description: The TCON TOP interface clock
+
+        clock-names:
+          minItems: 1
+          items:
+            - const: bus
+
+        clock-output-names: false
+        ports: false
+      required:
+        - power-domains
+
 examples:
   - |
     #include <dt-bindings/interrupt-controller/arm-gic.h>
-- 
2.53.0

